### PR TITLE
ci: Travis: use gcc9 with gcov job

### PR DESCRIPTION
**Pull request by blueyed**
- it is useful to test with the latest gcc in general
- it might help with gcov issues (merge mismatch)

### CODE REVIEW

**Comment by blueyed** [comment]
Fails:

> error: self-comparison always evaluates to true [-Werror=tautological-compare]

Could not reproduce this locally, so that might be something Arch's gcc has a patch for already.

Currently using `gcc-9 amd64 9.1.0-2ubuntu2~16.04`.

```
[139/262] Building C object src/nvim/CMakeFiles/nvim.dir/window.c.o
FAILED: /usr/lib/ccache/gcc-9 -DINCLUDE_GENERATED_DECLARATIONS -DMIN_LOG_LEVEL=3 -DNVIM_MSGPACK_HAS_FLOAT32 -DNVIM_UNIBI_HAS_VAR_FROM -DUSE_GCOV -D_GNU_SOURCE -Iconfig -I../src -Isrc/nvim/auto -Iinclude -I/home/travis/nvim-deps/usr/include/luajit-2.0 -isystem /home/travis/nvim-deps/usr/include --coverage -g   -Wall -Wextra -pedantic -Wno-unused-parameter -Wstrict-prototypes -std=gnu99 -Wshadow -Wconversion -Wmissing-prototypes -Wimplicit-fallthrough -Wvla -fstack-protector-strong -fdiagnostics-color=always -Wno-array-bounds -Werror  -Wno-conversion -MD -MT src/nvim/CMakeFiles/nvim.dir/window.c.o -MF src/nvim/CMakeFiles/nvim.dir/window.c.o.d -o src/nvim/CMakeFiles/nvim.dir/window.c.o   -c ../src/nvim/window.c
../src/nvim/window.c: In function ‘do_window’:
../src/nvim/window.c:183:33: error: self-comparison always evaluates to true [-Werror=tautological-compare]
  183 |     FOR_ALL_WINDOWS_IN_TAB(wp2, curtab) {
      |                                 ^~
../src/nvim/window.c: In function ‘win_valid’:
../src/nvim/window.c:1504:30: error: self-comparison always evaluates to true [-Werror=tautological-compare]
 1504 |   FOR_ALL_WINDOWS_IN_TAB(wp, curtab) {
      |                              ^~
../src/nvim/window.c: In function ‘win_count’:
../src/nvim/window.c:1536:30: Kerror: self-comparison always evaluates to true [-Werror=tautological-compare]
 1536 |   FOR_ALL_WINDOWS_IN_TAB(wp, curtab) {
      |                              ^~
../src/nvim/window.c: In function ‘one_window’:
../src/nvim/window.c:2244:30: error: self-comparison always evaluates to true [-Werror=tautological-compare]
 2244 |   FOR_ALL_WINDOWS_IN_TAB(wp, curtab) {
      |                              ^~
../src/nvim/window.c: In function ‘buf_jump_open_win’:
../src/nvim/window.c:4411:32: error: self-comparison always evaluates to true [-Werror=tautological-compare]
 4411 |     FOR_ALL_WINDOWS_IN_TAB(wp, curtab) {
      |                                ^~
../src/nvim/window.c: In function ‘win_size_save’:
../src/nvim/window.c:4749:30: error: self-comparison always evaluates to true [-Werror=tautological-compare]
 4749 |   FOR_ALL_WINDOWS_IN_TAB(wp, curtab) {
      |                              ^~
../src/nvim/window.c: In function ‘win_size_restore’:
../src/nvim/window.c:4768:34: error: self-comparison always evaluates to true [-Werror=tautological-compare]
 4768 |       FOR_ALL_WINDOWS_IN_TAB(wp, curtab) {
      |                                  ^~
../src/nvim/window.c: In function ‘win_setminheight’:
../src/nvim/window.c:5223:32: error: self-comparison always evaluates to true [-Werror=tautological-compare]
 5223 |     FOR_ALL_WINDOWS_IN_TAB(wp, curtab) {
      |                                ^~
../src/nvim/window.c: In function ‘only_one_window’:
../src/nvim/window.c:5995:30: error: self-comparison always evaluates to true [-Werror=tautological-compare]
 5995 |   FOR_ALL_WINDOWS_IN_TAB(wp, curtab) {
      |                              ^~
../src/nvim/window.c: In function ‘win_id2win’:
../src/nvim/window.c:6691:30: error: self-comparison always evaluates to true [-Werror=tautological-compare]
 6691 |   FOR_ALL_WINDOWS_IN_TAB(wp, curtab) {
      |                              ^~
cc1: all warnings being treated as errors
[139/262] Building C object src/nvim/C...o/msgpack_lua_c_bindings.generated.c.o
ninja: build stopped: subcommand failed.
```

**Comment by blueyed** [comment]
https://github.com/neovim/neovim/pull/10480 might fix it.
Not sure though why I am not seeing this myself - appears to be legit after all, isn't it?

**Comment by blueyed** [comment]
Now:
```
In file included from ../src/nvim/eval.c:6112:
src/nvim/auto/funcs.generated.h: In function ‘hash_internal_func_gperf’:
src/nvim/auto/funcs.generated.h:84:14: error: this statement may fall through [-Werror=implicit-fallthrough=]
   84 |         hval += asso_values[(unsigned char)str[13]];
      |         ~~~~~^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
src/nvim/auto/funcs.generated.h:86:7: note: here
   86 |       case 13:
      |       ^~~~
src/nvim/auto/funcs.generated.h:89:14: error: this statement may fall through [-Werror=implicit-fallthrough=]
   89 |         hval += asso_values[(unsigned char)str[10]];
      |         ~~~~~^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
src/nvim/auto/funcs.generated.h:91:7: note: here
   91 |       case 10:
      |       ^~~~
src/nvim/auto/funcs.generated.h:92:14: error: this statement may fall through [-Werror=implicit-fallthrough=]
   92 |         hval += asso_values[(unsigned char)str[9]];
      |         ~~~~~^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
src/nvim/auto/funcs.generated.h:94:7: note: here
   94 |       case 9:
      |       ^~~~
src/nvim/auto/funcs.generated.h:97:14: error: this statement may fall through [-Werror=implicit-fallthrough=]
   97 |         hval += asso_values[(unsigned char)str[6]];
      |         ~~~~~^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
src/nvim/auto/funcs.generated.h:99:7: note: here
   99 |       case 6:
      |       ^~~~
src/nvim/auto/funcs.generated.h:100:14: error: this statement may fall through [-Werror=implicit-fallthrough=]
  100 |         hval += asso_values[(unsigned char)str[5]];
      |         ~~~~~^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
src/nvim/auto/funcs.generated.h:102:7: note: here
  102 |       case 5:
      |       ^~~~
src/nvim/auto/funcs.generated.h:104:14: error: this statement may fall through [-Werror=implicit-fallthrough=]
  104 |         hval += asso_values[(unsigned char)str[3]];
      |         ~~~~~^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
src/nvim/auto/funcs.generated.h:106:7: note: here
  106 |       case 3:
      |       ^~~~
src/nvim/auto/funcs.generated.h:107:14: error: this statement may fall through [-Werror=implicit-fallthrough=]
  107 |         hval += asso_values[(unsigned char)str[2]];
      |         ~~~~~^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
src/nvim/auto/funcs.generated.h:109:7: note: here
  109 |       case 2:
      |       ^~~~
src/nvim/auto/funcs.generated.h:110:14: error: this statement may fall through [-Werror=implicit-fallthrough=]
  110 |         hval += asso_values[(unsigned char)str[1]];
      |         ~~~~~^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
src/nvim/auto/funcs.generated.h:112:7: note: here
  112 |       case 1:
      |       ^~~~
../src/nvim/eval.c: At top level:
cc1: error: unrecognized command line option ‘-Wno-static-in-inline’ [-Werror]
cc1: all warnings being treated as errors
[179/262] Building C object src/nvim/CMakeFiles/nvim.dir/ex_cmds.c.o
ninja: build stopped: subcommand failed.
+exit 1
The command "ci/script.sh" exited with 1.
```

**Comment by blueyed** [comment]
The above should be handled through the `PASSTHROUGH` comment in the file.. looks like `-Wimplicit-fallthrough=5` is used (requiring special attribute, not regexp comment).

**Comment by blueyed** [comment]
Looks like ccache is causing this?!

**Comment by blueyed** [comment]
Should be good/better after https://github.com/neovim/neovim/pull/10533.

File: ci/before_install.sh [hunk 1]
```diff
@@ -59,3 +59,16 @@ if [[ -n "$CMAKE_URL" ]]; then
...
+  set -ex
+  version=3.7.1
+  cd /tmp
+  curl -L --retry 5 --silent --fail "https://github.com/ccache/ccache/releases/download/v${version}/ccache-${version}.tar.bz2" | tar xj
+  cd "/tmp/ccache-${version}"
+  ./configure
+  make
+  sudo make install
+  ccache --version
+)
```
Comment (blueyed) [inline ci/before_install.sh]: Might be good to test latest ccache in one job (this one) still?
(similar to testing oldest CMake)

**Comment by blueyed** [comment]
> Might be good to test latest ccache in one job (this one) still?
> (similar to testing oldest CMake)

Pulled that out, but happy to create a separate PR if you find it useful.

File: .travis.yml [hunk 2]
```diff
@@ -46,15 +46,43 @@ env:
...
+      - pkg-config
+      - unzip
+      - valgrind
+      - xclip
+  homebrew:
+    update: true
+    packages:
+      - ccache
+      - ninja
+
```
Comment (blueyed) [inline .travis.yml]: This is only moved to allow using its new anchor below.

**Review by justinmk — APPROVED** [review]
Ok, but I hope we don't do more of this type of thing: It forces us to spend too much time managing compiler versions. Travis should be doing that for us.

Also, adding extra apt sources was a source of pain in the past.

**Comment by blueyed** [comment]
> It forces us to spend too much time managing compiler versions. Travis should be doing that for us.

Yes, and it still does.
It's just that it does not use the latest (nor oldest as with CMake), and it is good to have specific jobs for this then.

> Also, adding extra apt sources was a source of pain in the past.

I can imagine.  Therefore I think it is good to use YAML's features in this regard.  But I agree with your other comment that it should not become too complex.

**Pull request merged**